lib/cn/cn.c

```diff
@@ -1523,22 +1523,10 @@ cn_make(struct mpool *ds, const struct kvs_cparams *cp, struct kvdb_health *heal
     assert(cp);
     assert(health);
 
-    switch (cp->fanout) {
-        case 2:
-            fanout_bits = 1;
-            break;
-        case 4:
-            fanout_bits = 2;
-            break;
-        case 8:
-            fanout_bits = 3;
-            break;
-        case 16:
-            fanout_bits = 4;
-            break;
-        default:
-            return merr(EINVAL);
-    }
+    if (cp->fanout < CN_FANOUT_MIN || cp->fanout > CN_FANOUT_MAX)
+        return merr(EINVAL);
+
+    fanout_bits = ilog2(cp->fanout);
```

Comment:
IMO, better to just clamp it than to fail... thoughts?

Reply:
The check should never fail due to the params infrastructure. Just assert. Same think in cndb.c if statement.

Reply:
cn_make_test() circumvents your carefully laid plans..

Reply:
Then cn_make_test() is wrong.

Reply:
Please remove cn_make_test(). Not worth having.

Reply:
My take is that external module interfaces should guard against invalid inputs regardless of whether the build is debug or not, the exception being if they're performance sensitive then I'd use an assert/invariant.  cn_make() is not perf sensitive, so in this case see no reason to punish the caller with code that mysteriously doesn't work because you're not running a build with asserts enabled.

Reply:
I'd argue that most of our unit tests are garbage, but then they occassionally make a good catch.  Hard to figure out where to draw the line..

Reply:
Code organization bites us again...Ideally all cn parameters are exported from a libhse-cn. Annoying

Reply:
Wait cn_make() isn't even used outside of cn/... It should be put in a private header.

Reply:
Yeah, you missed the good ole days when for some reason we tried really hard for 100% code coverage which basically that meant that thousands of functions that should have been declared static were made public so that they could be fine-grain tested by some unit test that would basically fuzz the heck out of every function.  So now we have all these sub-module functions that are public which hurts the compiler's ability to optimize the code   LTO can figure it out, but that makes for a long running link phase..

Reply:
fuzzing functions has value.

Reply:
I liked what you had proposed to me earlier Alex of writing the test collection in the same file as the thing you are trying to test. Would solve a lot of problems. Would like to put together a plan for overhauling test organization and implementation at some point.

Reply:
seems like it should fail if `cp->fanout` is not a power of two.